solutions/cf-585D.mdx

```diff
@@ -0,0 +1,103 @@
+---
+id: cf-585D
+source: CF
+title: "Lizard Era: Beginning"
+author: Albert Ye
+---
+
+[Official Analysis](https://codeforces.com/blog/entry/20898)
+
+We notice that the size of the array is 25. A naive brute-force would take $\mathcal O(3^n)$ which is too slow because $n = 25$ and $3^25 = 8.47 \cdot 10^10 > 5 \cdot 10^9$. But, a meet-in-the-middle solution would take at most $\mathcal O(\sqrt{3^n} \log (\sqrt{3^n}))$, which is a significant optimization. 
```

Comment:
Prove the time complexity after the MIM-optimization... What makes it so fast? What did you optimize?